### ISSUE TITLE
[Bug] Qwen3-14B non-L3 decode_fwd generates repetitive text while L3 decode_all is correct

### ISSUE THREAD

**Issue by ndleslx** [hw-native-sys/pypto-lib#396]
### Diagnosis

**pypto-lib** — the issue is isolated to the optimized Qwen3-14B non-L3 `decode_fwd` layer implementation in `models/qwen3/14b/decode_layer.py`. The high-level serving loop, tokenizer, weights, RoPE config, page size, and L3 path all appear healthy. Replacing the `decode_fwd` layer body with the L3 `decode_all`-style implementation makes non-L3 generation correct.

Related: hw-native-sys/pypto-lib#362, hw-native-sys/pypto#1547

### Description

Qwen3-14B non-L3 serving generation diverges after the first decode steps and falls into repetitive output. The same prompt and serving setup generate good text through the L3 path.

Reproduction from `pypto-serving` with `pypto-lib` submodule at `3834f3d`:

```bash
cd /data/liuxu/pypto-serving
task-submit --device auto --max-time 0 --run \
  "PTO2_RING_HEAP=2147483648 PTO2_RING_TASK_WINDOW=262144 PTO2_RING_DEP_POOL=262144 \
   python examples/model/qwen3_14b/npu_generate.py \
     --model-dir /data/linyifan/models/Qwen3-14B \
     --prompt 'Huawei is' \
     --platform a2a3 \
     --max-seq-len 512 \
     --max-new-tokens 16"
```

Observed non-L3 output before the workaround:

```text
text:  a Chinese company. The company is located in the United States. The company is
token_ids: [264, 8453, 2813, 13, 576, 2813, 374, 7407, 304, 279, 3639, 4180, 13, 576, 2813, 374]
```

Control run with the L3 path:

```bash
task-submit --device auto --max-time 0 --run \
  "PTO2_RING_HEAP=2147483648 PTO2_RING_TASK_WINDOW=262144 PTO2_RING_DEP_POOL=262144 \
   python examples/model/qwen3_14b/npu_generate.py \
     --model-dir /data/linyifan/models/Qwen3-14B \
     --prompt 'Huawei is' \
     --platform a2a3 \
     --max-seq-len 512 \
     --max-new-tokens 64 \
     --l3"
```

Observed L3 output is fluent:

```text
text:  a Chinese multinational technology company that designs and sells consumer electronics, telecommunications equipment, and services. It is one of the world's largest and most valuable technology companies. Huawei was founded in 1987 by Ren Zhengfei, a former Chinese army engineer. The company has grown from a small trading company into a
```

Additional checks:

- Non-L3 1-token generation: ` a`
- Non-L3 2-token generation: ` a Chinese`
- Non-L3 diverges at token 3: non-L3 chooses `2813` (`company`), while L3 chooses `67926` (`multinational`).
- Added temporary validation to mark `k_cache` / `v_cache` as in/out tensors in `decode_fwd.py` and `prefill_fwd.py`.
- `prefill_fwd.py -p a2a3 -b 1 --num-layers 1 --seq-len 3` passes K/V/out, so short-prompt prefill tail handling is not the obvious cause.
- Full-depth prefill/decode show small K-cache/logit drift, but L3 generation remains good, so the serving failure is not explained by the high-level loop or generic numeric drift alone.

Workaround / isolating experiment:

1. Added an experimental `models/qwen3/14b/decode_layer_l3.py` with a `@pl.jit.inline` layer adapted from the L3 `qwen3_decode_all` body.
2. Changed `decode_fwd.py` to call `decode_layer_l3(...)` instead of the optimized `decode_layer(...)`.
3. Re-ran non-L3 generation.

Result with the L3-style decode layer inside non-L3 `decode_fwd`:

```text
text:  a Chinese multinational technology company that designs and sells consumer electronics, telecommunications equipment, and
token_ids: [264, 8453, 67926, 5440, 2813, 429, 14431, 323, 30778, 11502, 30491, 11, 61766, 7096, 11, 323]
```

A 64-token non-L3 run with the same workaround also remained fluent and matched the good L3 prefix.

This points to the current optimized `decode_layer.py` implementation, especially changes from the mixed cube/vector fused attention and scope-3 promotion path, rather than the serving dataflow.

### Environment

| Component | Version |
|---|---|
| pypto-lib | `4141b3f` (repo HEAD); reproduced in serving submodule at `3834f3d` |
| pypto | `a73c056a` (branch: detached) |
| pypto runtime (submodule) | `44005586` |
| pto-isa | `0a1ce522` (branch: detached) |
| ptoas | `ptoas 0.41` |
| CANN | `not detected` |

### Host Platform

Linux (aarch64)

### Additional Context

The direct suspect is the optimized non-L3 `decode_layer.py` path introduced around:

- `2d6f086` Qwen3-14B decode: mix-fuse cube+vec epilogues, promote scope-3 / scope-2 to flat pl.spmd (#387)
- Prior related decode changes: `640ce08`, `7bfe8d4`, `d439e1c`, `aacee12`

Suggested next debug step: compare top-k logits and per-layer hidden/attention outputs between current `decode_layer.py` and the L3 `qwen3_decode_all` body on the real serving tensors after the prefix `Huawei is a Chinese`.

**Comment by ndleslx** [comment]
Additional finding: the optimized non-L3 `decode_layer.py` appears to worsen decode correctness compared with the L3-style `decode_all` body.

Details:

- The current optimized non-L3 `decode_layer.py` can pass loose standalone logits validation, but serving generation diverges at token 3 for prompt `Huawei is`.
- Full-depth standalone cache-side validation shows small `k_cache` mismatches while `v_cache` passes.
- Replacing only the non-L3 layer body used by `decode_fwd.py` with an L3 `decode_all`-style layer restores good generation and matches the L3 prefix.
- This does not look like a page-size or KV address mapping error. It looks like the optimized mixed cube/vector fused attention or scope-3 optimized path introduces enough K-cache/hidden/logit drift to change greedy argmax.

So the practical regression is: optimized non-L3 `decode_layer.py` is less generation-correct than the L3 `decode_all` implementation, and the small K-cache drift is semantically significant for autoregressive generation.

**Comment by zhangqi-chen** [comment]
Closing as stale: no activity since 2026-05-28 (76 days), and the repro points at `models/qwen3/14b/decode_layer.py`, a path that no longer exists — `models/` was flattened to one directory per model build in #886 (2026-08-04) and the Qwen3 decode forward path has since been reworked (#670/#747, #756, #841). Please re-verify against current `main` and file a fresh bug if non-L3 decode still degenerates into repetitive output.
